source/distributed/grid_tools.cc

```diff
@@ -0,0 +1,184 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+
+
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/base/point.h>
+#include <deal.II/grid/tria.h>
+#include <deal.II/grid/grid_tools.h>
+#include <deal.II/distributed/tria.h>
+#include <deal.II/distributed/grid_tools.h>
+
+#ifdef DEAL_II_WITH_P4EST
+
+#include <vector>
+#include <cmath>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace parallel
+{
+  namespace GridTools
+  {
+    template < int dim, int spacedim>
+    std::vector<BoundingBox<spacedim> >
+    compute_locally_owned_bounding_box
+    (const parallel::distributed::Triangulation< dim, spacedim > &distributed_tria)
+    {
+      //Algorithm brief description: we want to build a number of BoundingBoxes
+      //enclosing all the locally owned cells.
+      //The idea is to find a group of parent cells which encloses all the locally
+      //owned cells and, for each of them, produce a bounding box of the locally owned
+      //points inside it.
+      //We chose, as group of cells, the children of the coarsest parent cell which contains
+      //both the first and the last locally owned cell in order to give a general
+      //description of the space and to describe the two connected region (if present)
+
+      // Begin by looking for those cells which we call firstc and lastc
+
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      firstc = distributed_tria.begin_active();
+      typename parallel::distributed::Triangulation< dim, spacedim >::active_cell_iterator
+      lastc = distributed_tria.last_active();
+
+      while ( ! (firstc->is_locally_owned() || firstc == lastc ) )
```

Comment:
I think the algorithm is general enough to accept a predicate instead of hard-coding locally owned cells

Reply:
Ok, then it would make sense to have it not only for distributed meshes...I see where this is going. Thank you!